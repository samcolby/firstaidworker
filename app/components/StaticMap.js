import React, { memo } from "react";
import PropTypes from "prop-types";

import {
  ActivityIndicator,
  Dimensions,
  StyleSheet,
  TouchableOpacity
} from "react-native";

import { Image } from "react-native-elements";

import { COLORS } from "../Constants";

import env from "../../env.json";

const STATIC_MAP_URL = "https://maps.googleapis.com/maps/api/staticmap";
const API_KEY = env.static_maps_key;

/**
 * React Pure Component to display
 * a static image of the person's location
 * @param {Object} props - Standard react props
 * @param {number} latitude
 *    The latitude of the center of the map
 * @param {number} longitude
 *    The longitude of the center of the map
 */
function StaticMap({ latitude, longitude, onPress }) {
  const { width } = Dimensions.get("window");

  const url = [
    STATIC_MAP_URL,
    `?center=${latitude},${longitude}`,
    "&zoom=16",
    `&size=${width}x${width}`,
    "&maptype=roadmap",
    `&markers=color:green%7C${latitude},${longitude}`,
    `&key=${API_KEY}`
  ].join("");

  return (
    <TouchableOpacity testID="staticMapLink" onPress={onPress}>
      <Image
        testID="staticMapImage"
        source={{ uri: url }}
        style={{ width: width, height: width, marginTop: 20 }}
        placeholderStyle={styles.placeholderStyle}
        PlaceholderContent={
          <ActivityIndicator size="large" color={COLORS.TAB_HINTS} />
        }
      />
    </TouchableOpacity>
  );
}

StaticMap.propTypes = {
  latitude: PropTypes.number.isRequired,
  longitude: PropTypes.number.isRequired,
  onPress: PropTypes.func
};

const styles = StyleSheet.create({
  placeholderStyle: { backgroundColor: COLORS.BACKGROUND }
});

export default memo(StaticMap);
