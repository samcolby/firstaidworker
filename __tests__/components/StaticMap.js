import React from "react";
import { render, fireEvent } from "react-native-testing-library";

import StaticMap from "../../app/components/StaticMap";

const testProps = {
  latitude: 51.2099548,
  longitude: -4.1149755,
  onPress: jest.fn()
};

describe("<StaticMap /> component", () => {
  it("renders correctly", () => {
    const { getByTestId, toJSON } = render(<StaticMap {...testProps} />);
    const component = getByTestId("staticMapLink");
    expect(component).toBeTruthy();

    expect(toJSON()).toMatchSnapshot();
  });

  it("displays the image centred on the correct latitude and longitude", () => {
    const { getByTestId } = render(<StaticMap {...testProps} />);
    const component = getByTestId("staticMapImage");

    const strRegExp =
      "center=" + testProps.latitude + "," + testProps.longitude;

    expect(JSON.stringify(component.props.source)).toMatch(
      new RegExp(strRegExp)
    );
  });

  it("triggers the onPress function", () => {
    const { getByTestId } = render(<StaticMap {...testProps} />);

    fireEvent(getByTestId("staticMapLink"), "press");

    expect(testProps.onPress).toHaveBeenCalled();
  });
});
