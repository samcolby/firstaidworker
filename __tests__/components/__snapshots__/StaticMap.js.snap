// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StaticMap /> component renders correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="staticMapLink"
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "position": "relative",
      }
    }
  >
    <Image
      onLoadEnd={[Function]}
      source={
        Object {
          "uri": "https://maps.googleapis.com/maps/api/staticmap?center=51.2099548,-4.1149755&zoom=16&size=750x750&maptype=roadmap&markers=color:green%7C51.2099548,-4.1149755&key=AIzaSyDYRidbnlD1wq1VJ45S7LxAteBk55tQnEU",
        }
      }
      style={
        Object {
          "height": 750,
          "marginTop": 20,
          "width": 750,
        }
      }
      testID="staticMapImage"
      theme={
        Object {
          "colors": Object {
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": Object {
              "android": Object {
                "error": "#f44336",
                "primary": "#2196f3",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "ios": Object {
                "error": "#ff3b30",
                "primary": "#007aff",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#8F0CE8",
            "success": "#52c41a",
            "warning": "#faad14",
          },
        }
      }
    />
    <View
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "height": 750,
            "justifyContent": "center",
            "marginTop": 20,
            "width": 750,
          }
        }
        testID="RNE__Image__placeholder"
      >
        <ActivityIndicator
          animating={true}
          color="red"
          hidesWhenStopped={true}
          size="large"
        />
      </View>
    </View>
  </View>
</View>
`;
